### DISCUSSION TITLE
PLC14500-Nano Loader (user interactive over serial port)

### DISCUSSION THREAD

**Discussion by wrh3c** [nicolacimmino/PLC-14500#5] [Show and tell]
I rewrote "bootloader.ino" to be interactive with the user over the serial port.  I call it "PLC14500-Nano Loader" (attached below).

This separates the functions of loading the program over serial (into Arduino memory), writing the program to Arduino EEPROM, and writing the program to PLC14500-Nano memory (board RAM), as well as viewing (as hex bytes or as disassembly) and editing the program in Arduino memory, and monitoring program execution.

 Primary features:
- Help message lists commands
- Input binary program from serial (works with Tera Term's "Send file..." with the Binary option checked)
- Load program from EEPROM, compare program to EEPROM, save program to EEPROM
- Write program to PLC14500-Nano board RAM
- Clear program, edit program (overwrite, insert, or delete, byte-by-byte), view program as hexadecimal bytes
- Disassemble program with or without hexadecimal bytes (output without hexadecimal bytes can be copied into an .asm file and re-assembled using the assembler!)
- Monitor address and data during program execution (not fast enough to keep up with full-speed clock, but useful in slow clock mode and single-step mode)
- Grab address and data from board (a way to capture the program from board RAM as it executes)

Enjoy!

[PLC14500_Nano_Loader_2023_03_12.zip](https://github.com/nicolacimmino/PLC-14500/files/10974627/PLC14500_Nano_Loader_2023_03_12.zip)

**Comment by wrh3c** [comment]
In case you're wondering about the 3-digit hexadecimal values in front of "I" (input binary program from serial), "L" (load program from EEPROM), and "S" (save program to EEPROM), I added them to prevent accidentally performing these operations by pressing the wrong key.  Here's how to remember them:
- 5EAI is SEriAl Input, where "S" is replaced with "5"
- 1EEL is Load EEprom Load, where the first "L" is replaced with "1"
- 5EES is Save EEprom Save, where the first "S" is replaced with "5"

The 3-digit hexadecimal values in front of "Z" (clear program) specify the method and value to use when clearing the program.  I arbitrarily chose the first digit as follows:
- "2" sets all values to the next 2 digits
- "3" sets first value, increments the value in each location by 1
- "4" sets first value, increments the high nybble (4 bits of data) first, then the low nybble (the 4-bit instruction code)

Clear program examples:

20FZ results in:
```
00: 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F
10: 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F
20: 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F
30: 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F
...
F0: 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F 0F
```

340Z results in:
```
00: 40 41 42 43 44 45 46 47 48 49 4A 4B 4C 4D 4E 4F
10: 50 51 52 53 54 55 56 57 58 59 5A 5B 5C 5D 5E 5F
20: 60 61 62 63 64 65 66 67 68 69 6A 6B 6C 6D 6E 6F
30: 70 71 72 73 74 75 76 77 78 79 7A 7B 7C 7D 7E 7F
...
F0: 30 31 32 33 34 35 36 37 38 39 3A 3B 3C 3D 3E 3F
```

400Z results in:
```
00: 00 10 20 30 40 50 60 70 80 90 A0 B0 C0 D0 E0 F0
10: 01 11 21 31 41 51 61 71 81 91 A1 B1 C1 D1 E1 F1
20: 02 12 22 32 42 52 62 72 82 92 A2 B2 C2 D2 E2 F2
30: 03 13 23 33 43 53 63 73 83 93 A3 B3 C3 D3 E3 F3
...
F0: 0F 1F 2F 3F 4F 5F 6F 7F 8F 9F AF BF CF DF EF FF
```
which disassembles to:
```
  00: 00  :  NOPO       ; no change in registers, set Flag O
  01: 10  :  NOPO 1     ; no change in registers, set Flag O
  02: 20  :  NOPO 2     ; no change in registers, set Flag O
  03: 30  :  NOPO 3     ; no change in registers, set Flag O
  04: 40  :  NOPO 4     ; no change in registers, set Flag O
  05: 50  :  NOPO 5     ; no change in registers, set Flag O
  06: 60  :  NOPO 6     ; no change in registers, set Flag O
  07: 70  :  NOPO 7     ; no change in registers, set Flag O
  08: 80  :  NOPO 8     ; no change in registers, set Flag O
  09: 90  :  NOPO 9     ; no change in registers, set Flag O
  0A: A0  :  NOPO 10    ; no change in registers, set Flag O
  0B: B0  :  NOPO 11    ; no change in registers, set Flag O
  0C: C0  :  NOPO 12    ; no change in registers, set Flag O
  0D: D0  :  NOPO 13    ; no change in registers, set Flag O
  0E: E0  :  NOPO 14    ; no change in registers, set Flag O
  0F: F0  :  NOPO 15    ; no change in registers, set Flag O
  10: 01  :  LD   SPR0  ; load SPR0
  11: 11  :  LD   SPR1  ; load SPR1
  12: 21  :  LD   SPR2  ; load SPR2
  13: 31  :  LD   SPR3  ; load SPR3
  14: 41  :  LD   SPR4  ; load SPR4
  15: 51  :  LD   SPR5  ; load SPR5
  16: 61  :  LD   SPR6  ; load SPR6
  17: 71  :  LD   SPR7  ; load SPR7
  18: 81  :  LD   IN0   ; load IN0
  19: 91  :  LD   IN1   ; load IN1
  1A: A1  :  LD   IN2   ; load IN2
  1B: B1  :  LD   IN3   ; load IN3
  1C: C1  :  LD   IN4   ; load IN4
  1D: D1  :  LD   IN5   ; load IN5
  1E: E1  :  LD   IN6   ; load IN6
  1F: F1  :  LD   IN7   ; load IN7
  20: 02  :  LDC  SPR0  ; load complement of SPR0
  21: 12  :  LDC  SPR1  ; load complement of SPR1
  22: 22  :  LDC  SPR2  ; load complement of SPR2
  23: 32  :  LDC  SPR3  ; load complement of SPR3
  ...
  F8: 8F  :  NOPF 8     ; no change in registers, set Flag F
  F9: 9F  :  NOPF 9     ; no change in registers, set Flag F
  FA: AF  :  NOPF 10    ; no change in registers, set Flag F
  FB: BF  :  NOPF 11    ; no change in registers, set Flag F
  FC: CF  :  NOPF 12    ; no change in registers, set Flag F
  FD: DF  :  NOPF 13    ; no change in registers, set Flag F
  FE: EF  :  NOPF 14    ; no change in registers, set Flag F
  FF: FF  :  NOPF 15    ; no change in registers, set Flag F
```

**Comment by nicolacimmino** [comment]
I am also actually in the process of adding an inteactive inteface to the bootloader. Probably because I grew up with a Commodore 64 I went for a classic asm/monitor interface. 

I like the monitor address/data. Gave me the idea to add something similar, it could show the current address/data in bin/hex and decode the instruction too. Would work as a great aid while stepping through a program. I'll make sure I credit your idea when I get to it.

**Reply by wrh3c** [reply]
Sounds good!  I appreciate you including the bootloader code so I could build on top of it.

Give my code (attached in the original discussion) a try - monitor mode currently shows the address, data, and decoded (disassembled) instruction!
